src/System.Buffers/src/System/Buffers/BufferBucket.cs

```diff
@@ -0,0 +1,82 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Diagnostics;
+using System.Threading;
+
+namespace System.Buffers
+{
+    internal sealed class BufferBucket
+    {
+        private volatile int _index;
+        private byte[][] _data;
+        private int _bufferSize;
+        private SpinLock _lock;
+
+        internal BufferBucket(int bufferSize, int numberOfBuffers)
+        {
+            _index = 0;
+            _bufferSize = bufferSize;
+            _lock = new SpinLock();
+            _data = new byte[numberOfBuffers][];
+            for (int i = 0; i < numberOfBuffers; i++)
+            {
+                _data[i] = new byte[bufferSize];
+            }
+        }
+
+        internal byte[] Rent()
+        {
+            byte[] buffer;
+
+            // Use a SpinLock since it is super lightweight
+            // and our lock is very short lived
+            bool taken = false;
+            _lock.Enter(ref taken);
+            Debug.Assert(taken);
+
+            // Be quick and do an unsynchronized read to see if all
+            // of our buffers have been used
+            if (_index + 1 >= _data.Length)
```

Comment:
Why `+ 1` and `>=`, subtract one from both sides and get `if (_index > _data.Length)`, though since it's about to be a deref maybe you still want `>=`...